### PR TITLE
[WIP] Update documentation of "astropy.wcs"

### LINKED ISSUE #9748: Update documentation of astropy.wcs
The `astropy.wcs` documentation has not been updated for a while.
I'm opening this issue to serve as a place holder for tasks to do related to improving the documentation in 4.1.

[ ] Write examples using the WCS API.

### PR DESCRIPTION
**Pull request by kakirastern**
### Description

This pull request is to update the documentation of `astropy.wcs`, as suggested in issue #9748 by @nden. The main page of the new documentation will have the following structure: 

1. Introduction
2. A Simple Example
    * Load from a file 
    * Use the high level interface to transform between pixels and world
3. Using WCS (links to other pages, similar to `modelling`'s docs) 
    * High level interface in detail (including examples)
    * Low level interface
    * Using the core wcslib transforms
    * Example creating a WCS programmatically
        i. WCS object as defined programmatically with the use of a Python dictionary
        ii. WCS (objects) for 2D and 3D data and beyond
    * Supported projections (with example CTYPE keywords)
    * A section on WCS tools
    * Relax constants - a summary: a table with one row for each one with a short description

Fixes #9748

### CODE REVIEW

**Comment by nden** [comment]
tagging also @perrygreenfield @mcara @shannnnnyyy @hcferguson @eslavich

**Comment by pllim** [comment]
Test failure is related. Please run tests locally before pushing commits.

**Note for reviewers:** When this is ready for copyediting, please ping the copyeditor for review.

**Comment by nden** [comment]
@pllim  Using the astropy infrastructure to do this rewrite of the documentation would be much simpler than providing a ready to merge PR. It's a **WIP PR** and we know the failure is related. As we work through the plan provided above there will be other commits which are incomplete.

 Do you object because there will be a few more runs of Travis? We could try to limit the test runs. Can you clarify your comment ` Please run tests locally before pushing commits`? as it is not possible to do this in this case.

**Comment by pllim** [comment]
@nden , please clarify what is not possible. I think a local run (not even a full run, but just a run touching the `wcs` stuff) would have caught:

```
021     >>> hdu = fits.open('/Users/username/some_data_file.fits')
UNEXPECTED EXCEPTION: FileNotFoundError(2, 'No such file or directory')
```

Testing locally is recommended as per https://astropy.readthedocs.io/en/latest/development/workflow/development_workflow.html#the-editing-workflow

**Comment by kakirastern** [comment]
Thanks @pllim. I just hesitated to add `# doctest: +IGNORE_WARNINGS` to the new example above like the old example below. My apologies.

**Comment by pllim** [comment]
No worries. The problem is that your example is trying to access a non-existent file. Either skip it or use an existing test file. Thanks for your patience and contributions!

**Comment by nden** [comment]
I expect to have other similar cases until this PR is done. Some of the sections are placeholders.

**Comment by bsipocz** [comment]
> I expect to have other similar cases until this PR is done. Some of the sections are placeholders.

Doing WIP commits with `[skip ci]` would be a perfect workround for that

**Comment by pllim** [comment]
Re: placeholders -- FWIW, adding a `#doctest: +SKIP` directive to them would at least make it possible to test the rest of the stuff. If you are not ready to run the CI at all, can also add `[ci skip]` to the commit message. FYI.

Having too many CI runs may also hold up CI of other PRs (things get queue up when there is high traffic). So, please use your good judgement. Thanks.

File: docs/wcs/index.rst [hunk 1]
```diff
@@ -11,248 +11,114 @@ World Coordinate System (`astropy.wcs`)
...
-
-.. note::
-    The members of the WCS object correspond roughly to the key/value
-    pairs in the FITS header.  However, they are adjusted and
-    normalized in a number of ways that make performing the WCS
-    transformation easier.  Therefore, they can not be relied upon to
-    get the original values in the header.  To build up a FITS header
-    directly and specifically, use `astropy.io.fits.Header` directly.
-
-.. _wcslint:
```
Comment (pllim) [inline docs/wcs/index.rst]: Removing this section ref failed the `html-docs` build. FYI.

**Comment by kakirastern** [comment]
> > I expect to have other similar cases until this PR is done. Some of the sections are placeholders.
> 
> Doing WIP commits with `[skip ci]` would be a perfect workaround for that

Thanks @bsipocz I will remember to do that next time

**Comment by pllim** [comment]
If you simply want to see the rendered HTML pages, you can also put `[skip travis]` (not `[ci skip]`) in the commit message. That will skip Travis CI but allows CircleCI to run and general HTML, which will only work if docs can build successfully in the first place.

I know this is a lot of info at once, so feel free to reach out to us if you need help or clarifications. Thanks again!

File: docs/wcs/index.rst (hunk 1 above)
Comment (kakirastern) [reply docs/wcs/index.rst]: Thanks @pllim for the heads-up.

**Comment by kakirastern** [comment]
Have just updated the FITS file used in the example used in the "A Simple Example" section to a Multi-Extension FITS format with more than 1 extension as suggested by @nden earlier, using the file at `astropy/wcs/tests/data/j94f05bgq_flt.fits` instead.

**Comment by kakirastern** [comment]
Merged introduction into the PR.

File: docs/wcs/index.rst (hunk 1 above)
Comment (kakirastern) [reply docs/wcs/index.rst:130]: Section restored at least temporarily...

File: docs/wcs/index.rst [hunk 2]
```diff
@@ -11,13 +11,27 @@ World Coordinate System (`astropy.wcs`)
...
+(WCS) transformations defined in several elaborate `FITS WCS standard`_ conventions.
+These transformations work both forward (from pixel to world) and backward
+(from world to pixel).
+
+For historical reasons and to support legacy software, `astropy.wcs` maintains
+two separate application interfaces. The ``High-Level API`` should be used by
+most applications. It abstracts out the underlying object and works transparently
+with other packages which support the
+`Common Python Interface for WCS <https://zenodo.org/record/1188875#.XnpOtJNKjyI>`_,
+ allowing for a more flexible approach to the problem and avoiding the `limitations
```
Comment (nden) [inline docs/wcs/index.rst:29]: It looks like the indentation starting here is causing the html build to fail

Comment (kakirastern) [reply docs/wcs/index.rst]: No problem, just changed the indentation to fix.

**Comment by kakirastern** [comment]
Added "Supported projections" section to the new `wcs` docs using old content(s). But am wondering if the details regarding the versions of `wcslib` package need to be updated. Can add more info to the subsection once confirmed.

**Comment by kakirastern** [comment]
Just added the `Loading WCS information from a FITS file` subsection with old content(s) as well. May need to update the example for this subsection separately in `examples/from_file.py` sometime later.

File: docs/wcs/index.rst [hunk 3]
```diff
@@ -28,89 +42,139 @@ It performs three separate classes of WCS transformations:
...
+    CTYPE : 'RA---TAN-SIP'  'DEC--TAN-SIP'
+    CRVAL : 5.63056810618  -72.05457184278998
+    CRPIX : 2048.0  1024.0
+    CD1_1 CD1_2  : 1.29056256197165e-05  5.95309123310338e-06
+    CD2_1 CD2_2  : 5.0220581265601e-06  -1.2644774105568e-05
+    NAXIS : 1  1
+
+It is a good habit to check how many extensions are there, in case the FITS file is
+of the multi-extension type.
+
```
Comment (nden) [inline docs/wcs/index.rst:102]: I think the `Simlpe example` section should use the high level interface. All we need is a call to `wcs.pixel_to_world` and a call to `wcs.world_to_pixel`, and printing the results.

The current example should go under the `Low Level Interface` section.

File: docs/wcs/index.rst [hunk 4]
```diff
@@ -28,89 +42,139 @@ It performs three separate classes of WCS transformations:
...
 .. literalinclude:: examples/from_file.py
    :language: python
 
-Building a WCS structure programmatically
------------------------------------------
+Using `astropy.wcs`
+===================
+
+High level interface
+--------------------
```
Comment (nden) [inline docs/wcs/index.rst:168]: Please rename this section to `Shared Python Interface for World Coordinate Systems` and  link it to wcsapi.rst.

File: docs/wcs/index.rst [hunk 5]
```diff
@@ -28,89 +42,139 @@ It performs three separate classes of WCS transformations:
...
+
+Low level interface
+-------------------
```
Comment (nden) [inline docs/wcs/index.rst:171]: The rename this section to `Legacy Interface` and move the current `Simple example` here.

File: docs/wcs/index.rst [hunk 6]
```diff
@@ -28,89 +42,139 @@ It performs three separate classes of WCS transformations:
...
+
+Using the core wcslib transforms
+--------------------------------
```
Comment (nden) [inline docs/wcs/index.rst:174]: This should go as a subsection of `Legacy Interface` with an example of the wcs.wcs_* methods.

File: docs/wcs/index.rst (hunk 4 above)
Comment (nden) [reply docs/wcs/index.rst:168]: wcs.api should be removed from `Other information`

File: docs/wcs/index.rst [hunk 7]
```diff
@@ -129,41 +193,8 @@ saves those settings to a new FITS header.
...
-        - 'unitfix' made the change 'Changed units: 'HZ      ' -> 'Hz''.
-
-Bounds checking
----------------
-
-Bounds checking is enabled by default, and any computed world
-coordinates outside of [-180°, 180°] for longitude and [-90°, 90°] in
-latitude are marked as invalid.  To disable this behavior, use
-`astropy.wcs.Wcsprm.bounds_check`.
-
```
Comment (nden) [inline docs/wcs/index.rst:163]: Please add `Validation and Bounds checking` to `Other information`. I guess it needs to be a separate rst file.

File: docs/wcs/index.rst [hunk 8]
```diff
@@ -211,48 +242,11 @@ distortions are supported:
...
-included.
-
-The pixel scales of a celestial image or the pixel dimensions of a non-celestial
-image can be extracted with the utility functions
-`~astropy.wcs.utils.proj_plane_pixel_scales` and
-`~astropy.wcs.utils.non_celestial_pixel_scales`. Likewise, celestial pixel
-area can be extracted with the utility function
-`~astropy.wcs.utils.proj_plane_pixel_area`.
+WCS tools
+---------
```
Comment (nden) [inline docs/wcs/index.rst:246]: Link `WCS tools` this to a separate rst file. Add the `Subsetting and Pixel Scales` section above to that file.
Add also `Matplotlib plots with correct WCS projection` as a subsection to that file.

File: docs/wcs/index.rst [hunk 9]
```diff
@@ -28,89 +42,139 @@ It performs three separate classes of WCS transformations:
...
+--------------------
+
+Low level interface
+-------------------
+
+Using the core wcslib transforms
+--------------------------------
+
+Examples creating a WCS programmatically
+----------------------------------------
```
Comment (nden) [inline docs/wcs/index.rst:177]: This section should include several examples, ideally each of them in a separate file and linked here, something like

.. toctree::
   :maxdepth: 2

   Create an imaging WCS <example_create_imaging>
   Create a cube WCS <example_cube_wcs>

where the first example is the existing one in the current documentation and the second is the one you suggested from sunpy.

File: docs/wcs/index.rst (hunk 3 above)
Comment (kakirastern) [reply docs/wcs/index.rst:88]: Yup, have added one example for `wcs. pixel_to_world` and another one for `wcs.world_to_pixel` in the "A Simple Example" section. (Maybe I should rename this to "Simple Examples" next time?)

File: docs/wcs/index.rst (hunk 6 above)
Comment (kakirastern) [reply docs/wcs/index.rst]: Have added one example for `wcs.wcs_pix2world` and one for `wcs.wcs_world2pix` in the "Using the core wcslib transforms" subsection.

**Comment by kakirastern** [comment]
Apologies for the many tries and CI tests ran to get the CI checks passed on GitHub, esp. the `ci/cicleci: html-docs` test. I have tried testing locally for the `docs_build` but since I am using a computer with macOS, I got the case bug regarding `m_bol`/`M_bol` docs, so am stuck locally. Will try my best to restrict the number of such instances though.

**Comment by kakirastern** [comment]
Looks like the `ci/circleci: html-docs` failed due to some server-side issue unrelated to this PR:
```
Warning, treated as error:
failed to reach any of the inventories with the following issues:
intersphinx inventory 'http://docs.astropy.org/projects/package-template/en/latest/objects.inv' not fetchable due to <class 'requests.exceptions.HTTPError'>: 502 Server Error: Bad Gateway for url: http://docs.astropy.org/projects/package-template/en/latest/objects.inv
ERROR: InvocationError for command /home/circleci/project/.tox/build_docs/bin/sphinx-build -W -b html . _build/html (exited with code 2)
build_docs: Shutting down tox-pypi-filter server
___________________________________ summary ____________________________________
ERROR:   build_docs: commands failed
```

**Comment by kakirastern** [comment]
> @kakirastern I left inline suggestions on how to proceed. I hope they are clear but let me know if not :-).

@nden I have made some changes according to the inline suggestions you have provided. They are very clear! 😃Please let me know if further changes need be made. Looks like the CI tests should be passing now except for the server-side error as documented above.

**Comment by mhvk** [comment]
Removed the 4.1 milestone since not release critical. Though very little is better than improved documentation, so let's be sure this gets in for 4.2!

File: docs/wcs/index.rst [hunk 10]
```diff
@@ -28,79 +42,103 @@ It performs three separate classes of WCS transformations:
...
+indicating whether the input arrays are 0- or 1-based.
+The Low-level interface assumes Cartesian order (x, y) of the input coordinates,
+however the Common Interface for World Coordinate System accepts both conventions.
+The order of the pixel coordinates ((x, y) vs (row, column)) in the Common API
+depends on the method or property used, and this can normally be determined from
+the property or method name. Properties and methods containing “pixel” assume (x, y)
+ordering, while properties and methods containing “array” assume (row, column) ordering.
 
-    1. ``from astropy import wcs``
+Some Simple Examples
```
Comment (nden) [inline docs/wcs/index.rst:64]: Let's leave just one example in this section. 
I would call it `A Simple Example` and use one of the data files in the test subdirectory of the wcs package (see below). It needs to convert between detector and sky using the high level interface.

File: docs/wcs/index.rst [hunk 11]
```diff
@@ -28,79 +42,103 @@ It performs three separate classes of WCS transformations:
...
-       `~astropy.io.fits.HDUList` object.
+One example of the use of the high-level WCS API is to use the
+`~astropy.wcs.wcs.WCS.pixel_to_world` to yield the simplest WCS
+with default values, converting from pixel to world coordinates::
 
-    3. Optionally, if the FITS file uses any deprecated or
-       non-standard features, you may need to call one of the
-       `~astropy.wcs.wcs.WCS.fix` methods on the object.
+    >>> from astropy.wcs import WCS
+    >>> w = WCS('image.fits')  # doctest: +IGNORE_WARNINGS
```
Comment (nden) [inline docs/wcs/index.rst:72]: It'd be good to use a multiextension FITS file to show that the header of the extension with the WCS keywords should be passed to the constructor. For example

```
from astropy.io import fits
from astropy import wcs

f = get_pkg_data_filename('data/j94f05bgq_flt.fits', package='astropy.wcs.tests')
w = WCS(f[1].header)
```
I think this should also pass a test so no need to ignore them.

File: docs/wcs/index.rst [hunk 12]
```diff
@@ -28,79 +42,103 @@ It performs three separate classes of WCS transformations:
...
+    >>> lon, lat = w.pixel_to_world(30, 40)
```
Comment (nden) [inline docs/wcs/index.rst:73]: This method returns a `SkyCoord` object and I want to show this. I would change the example to

```
sky = w.pixel_to_world(30, 40)
print(sky)
<SkyCoord (ICRS): (ra, dec) in deg
    (5.52844243, -72.05207809)>
```

File: docs/wcs/index.rst [hunk 13]
```diff
@@ -28,79 +42,103 @@ It performs three separate classes of WCS transformations:
...
+    >>> print(lon, lat)
+    31.0 41.0
 
-    4. Use one of the following transformation methods:
+Similarly, another use of the high-level API is to use the
+`~astropy.wcs.wcs.WCS.world_to_pixel` to yield another simple WCS, while
+converting from world to pixel coordinates::
```
Comment (nden) [inline docs/wcs/index.rst:79]: This method transforms world to pixel coordinates .

Rewrite the example below as 

>>> x, y = w.pixel_to_world(sky)
>>> print(x, y)

File: docs/wcs/index.rst [hunk 14]
```diff
@@ -28,79 +42,103 @@ It performs three separate classes of WCS transformations:
...
 
-       - From pixels to world coordinates:
+    >>> from astropy.wcs import WCS
+    >>> from astropy.units import Unit
+    >>> w = WCS('image.fits')  # doctest: +IGNORE_WARNINGS
+    >>> lon, lat = w.world_to_pixel(30*Unit(''), 40*Unit(''))
+    >>> print(lon, lat)
+    29.0 39.0
+
+It should be noted the following transformations can be applied:
```
Comment (nden) [inline docs/wcs/index.rst:88]: Everything starting here ... and until `Using the WCS` section should be removed from this page.
The examples with other methods should go under `Low Level Interface` under `Using WCS`.

File: docs/wcs/index.rst [hunk 15]
```diff
@@ -109,60 +147,95 @@ the WCS cards in its primary header to transform.
...
+    Number of WCS axes: 2
+    CTYPE : 'RA---TAN-SIP'  'DEC--TAN-SIP'
+    CRVAL : 5.63056810618  -72.05457184278998
+    CRPIX : 2048.0  1024.0
+    CD1_1 CD1_2  : 1.29056256197165e-05  5.95309123310338e-06
+    CD2_1 CD2_2  : 5.0220581265601e-06  -1.2644774105568e-05
+    NAXIS : 1  1
+
+It is a good habit to check how many extensions are there, in case the FITS file is
+of the multi-extension type.
```
Comment (nden) [inline docs/wcs/index.rst:197]: Ah, this is actually the example I was thinking of putting under `A Simple Example` above (together with the two calls  `pixel_to_world` and `world_to_pixel`. 

And move all other example above here, under  the `Legacy Interface` section . Then move the text in `Legacy Interface` to a separate file and link to it in the `Using astropy.wcs` section.

**Comment by kakirastern** [comment]
Not a problem @nden I will work on this PR a bit more over the next few days

**Comment by kakirastern** [comment]
@nden Have made the changes you have suggested. Hopefully I have not misunderstood anything and added to your burden inadvertently. 😊 Please let me know of any new changes needed next.

File: docs/wcs/index.rst (hunk 11 above)
Comment (kakirastern) [reply docs/wcs/index.rst]: Sure

**Comment by kakirastern** [comment]
The `html-docs` test bug seems really weird: 
```
Warning, treated as error:
/home/circleci/project/.tox/build_docs/lib/python3.7/site-packages/astropy/wcs/wcs.py:docstring of astropy.wcs.WCS:1:py:class reference target not found: astropy.wcs.wcsapi.fitswcs.FITSWCSAPIMixin
ERROR: InvocationError for command /home/circleci/project/.tox/build_docs/bin/sphinx-build -W -b html . _build/html (exited with code 2)
build_docs: Shutting down tox-pypi-filter server
```

It doesn't seem to be directly related to this PR, which concerns with the docs only and not code.

**Comment by pllim** [comment]
Re: Weird failure -- Have you rebased against upstream master lately?

**Comment by kakirastern** [comment]
@pllim I did `git pull origin master`, but I could do a rebase soon

**Comment by pllim** [comment]
`git pull origin master` sounds like you are pulling your own fork's `master` branch?

**Comment by pllim** [comment]
Try:
```
git fetch upstream master
git rebase upstream/master
```

If needed, rename `upstream` to whatever you name the remote that points back to this repo, not your fork. If there are conflicts, you need to resolve them, but based on the status reported here, I don't think there will be. After you rebase, you need to force push, so make a backup branch before rebase if you are not sure. Hope this helps!

**Comment by kakirastern** [comment]
Okay, sure! Let me try it. Just force pushed after rebasing.

**Comment by nden** [comment]
@kakirastern This looks now much better. Could you rebase against master. I made a small further reorg on a separate branch and I want to create a PR against your branch after you rebase. Thanks!

**Comment by kakirastern** [comment]
Hi @nden I have just rebased... Seems to have some problem with passing the `ci/circleci: html-docs` tests though.

**Comment by kakirastern** [comment]
Sure @nden, am working on fixing the build problems again. Will let you know if I run into any issues. 😁

**Comment by kakirastern** [comment]
Hopefully we can get all the CI tests passed soon 🤞🏼

**Comment by pllim** [comment]
If you see some weird `__call__` failure with Sphinx 3.1.0, you might need to *rebase* against `master` to pick up the fix.

You might have to *rebase* anyway to get rid of those merge commits you have. FYI.

**Comment by kakirastern** [comment]
Thanks for the reminder @pllim!

**Comment by nden** [comment]
Thank you @kakirastern !

**Comment by kakirastern** [comment]
You are welcome @nden And, the `ci/circleci: html-docs` test is passing finally!

**Comment by kakirastern** [comment]
Looks like all CI tests are passing now. PR is ready for a review.

**Comment by nden** [comment]
Awesome, thanks @kakirastern! Tagging a few people here that may want to review. I think this sets the new overall look of the documentation. I suggest we merge this PR soon and improve in smaller PRs.
@astrofrog @Cadair @mcara @perrygreenfield

**Comment by nden** [comment]
@kakirastern Please squash all commits and then I think it's ready to merge.

**Comment by kakirastern** [comment]
@nden I have just rebased and squashed all commits into one.

**Comment by kakirastern** [comment]
You are welcome @nden! It has been my pleasure

**Comment by mcara** [comment]
@kakirastern Thank you

**Comment by pllim** [comment]
I didn't realize `wcslint` doc was completely removed int his patch. Does anyone remember if it was intentional?

**Comment by mcara** [comment]
@pllim it seems like the content of was spread over several new files such as:

[docs/wcs/supported_projections.rst](https://github.com/astropy/astropy/blob/ec194a78a254fa73dec0a1fd83320780b8193578/docs/wcs/supported_projections.rst)
[docs/wcs/validation.rst](https://github.com/astropy/astropy/blob/ec194a78a254fa73dec0a1fd83320780b8193578/docs/wcs/validation.rst)

and probably the `wcslint` section was accidentally removed in this process - https://github.com/astropy/astropy/pull/10058/files#r396526059

**Comment by pllim** [comment]
@mcara , the original PR author is no longer active. So who should add this missing doc back and where?

**Comment by mcara** [comment]
I would put it back where it was or make a separate `wcslint.rst` file if you wish

**Comment by pllim** [comment]
I'll leave where it goes in your good judgement, as long as `.. _wcslint:` goes with it. Thanks! 🙏

**Pull request merged**